pilot/utils/task.py

```diff
@@ -0,0 +1,147 @@
+import json
+from uuid import uuid4
+
+from utils.telemetry import telemetry
+from utils.exit import trace_code_event
+from const.telemetry import LOOP_THRESHOLD
+
+
+class Task:
+    """
+    Task data structure to store information about the current task. The task data structure is sent to telemetry.
+    Currently used to trace big loops in the code.
+
+    >>> from utils.task import Task
+
+    To set up a new task:
+
+    >>> task = Task()
+
+    To set a value:
+
+    >>> task.set('task_description', 'test')
+
+    To increment a value:
+
+    >>> task.inc('steps')
+
+    To start a new task:
+
+    >>> task.start_new_task('test', 1)
+
+    When debugging recursion happens inside a task (see pilot/helpers/Debugger.py) we add a debugging task to the
+    task data structure. To add a debugging task:
+
+    >>> task.add_debugging_task(1, {'command': 'test'}, 'This is not working', 'Command is not working')
+
+    To clear the task:
+
+    >>> task.clear()
+
+    To send the task:
+
+    >>> task.send()
+
+    Note: the task will be sent automatically if the number of steps exceeds the threshold
+    """
+
+    def __init__(self):
+        self.initial_data = {
+            'task_description': '',
+            'task_number': 0,
+            'steps': 0,
+            'debugging': [],
+        }
+        self.data = self.initial_data.copy()
+        self.ping_extension = True
+
+    def set(self, key: str, value: any):
+        """
+        Set a value in the task data
+
+        :param key: key to set
+        :param value: value to set
+        """
+        self.data[key] = value
+
+    def inc(self, key: str, value: int = 1):
+        """
+        Increment a value in the task data
+
+        :param key: key to increment
+        :param value: value to increment by
+        """
+        self.data[key] += value
+        if key == 'steps' and self.data[key] == LOOP_THRESHOLD + 1:
```

Comment:
Looks like this would send telemetry for *every* task in that loop over the threshold, instead of just the first one?